worker/provisioner/broker.go

```diff
@@ -118,6 +118,26 @@ func finishNetworkConfig(bridgeDevice string, interfaces []network.InterfaceInfo
 	return results, nil
 }
 
+// findDNSServerConfig is a heuristic method to find an adequate DNS
+// configuration. Currently the only rule that is implemented is that common
+// configuration files are parsed until a configuration is found that is not a
+// loopback address (i.e systemd/resolved stub address).
+func findDNSServerConfig() (*network.DNSConfig, error) {
+	for _, dnsConfigFile := range resolvConfFiles {
+		dnsConfig, err := network.ParseResolvConf(dnsConfigFile)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		for _, nameServer := range dnsConfig.Nameservers {
+			if nameServer.Scope != network.ScopeMachineLocal {
+				logger.Debugf("The DNS configuration from %s has been selected for use", dnsConfigFile)
+				return dnsConfig, nil
```

Comment:
this seems to say that if any nameserver we find isn't local then we return all nameservers. Would it be better to always strip machine local and then return only if we have a list that is nonempty?

Note that parsing resolve.conf was always more of a hack than good process. So I'd prefer if we trusted the config from MAAS for the container over happenstance parsing of the host machine.

I'm ok with extending the parsing. But what we really want is something more like "in this space/subnet, this is the nameserver and search path to use" coming from MAAS.

Note that /etc/netplan/99-juju.yaml on the host machine has the right nameserver configured. So clearly somewhere we really do have the right values from an authoritative source.

Reply:
> this seems to say that if any nameserver we find isn't local then we return all nameservers. Would it be better to always strip machine local and then return only if we have a list that is nonempty?

Acknowledged, that is better.

> Note that parsing resolve.conf was always more of a hack than good process. So I'd prefer if we trusted the config from MAAS for the container over happenstance parsing of the host machine.

> I'm ok with extending the parsing. But what we really want is something more like "in this space/subnet, this is the nameserver and search path to use" coming from MAAS.

I agree, but this portion of the code is the "Do this only If provider did not give us any DNS settings" part of the code. Which means what you are suggesting would go in the environment specific portion of the code (for MAAS) which I believe at the moment isn't even attempting to capture any DNS configuration info. MAAS2 gets the DNS info [here](https://github.com/ExternalReality/juju/blob/348e2b44280e6e931e92aa9c93a5067b8531964e/provider/maas/interfaces.go#L231) but fails to provide it to the container (although it does look like there is that looks like it is trying to provide that info).
https://github.com/ExternalReality/juju/blob/348e2b44280e6e931e92aa9c93a5067b8531964e/provider/maas/interfaces.go#L231

I'll quickly check a few more things to see if I cannot just simply get to provider to supply the correct information rather than increase the complexity of the fallback (which is what this patch does.)

I think another alternative (for the fallback) would be to ask if the init system is systemd and if so we can ask systemd directly what it thinks the appropriate DNS config should be. There already exists init system discovery code (for agent installation) which makes use of the coreos systemd golang library which is already a dependency.